### PR TITLE
Environ utils

### PR DESCRIPTION
**Pull request by vickumar1981**
@marcesher @sleitner  @imuchnik 

Changes to environ utils as discussed on previous PR #31.

env is now a member variable of the EnvironUtils class.  The isDev() and isProd() functions are non-static.

### CODE REVIEW

File: src/main/groovy/jenkins/automation/utils/EnvironmentUtils.groovy [hunk 1]
```diff
@@ -7,24 +7,26 @@ package jenkins.automation.utils
...
  *  relies on ENVIRONMENT env variable in Jenkins
  */
 
-    static isDev(String env) {
-        return env.toUpperCase() as Environment== Environment.DEV
+    String env
+
+    boolean isDev() {
+        return env.toLowerCase() as Environment== Environment.dev
     }
```
Comment (sleitner) [inline src/main/groovy/jenkins/automation/utils/EnvironmentUtils.groovy:33]: isStage?

Comment (vickumar1981) [inline src/main/groovy/jenkins/automation/utils/EnvironmentUtils.groovy:33]: doesn't exist.  let me add it.

Comment (imuchnik) [inline src/main/groovy/jenkins/automation/utils/EnvironmentUtils.groovy:33]: Those were arbitrarily names. We will need to still talk about it.

**Comment by sleitner** [comment]
:+1:

**Comment by imuchnik** [comment]
:-1: Making the class non-static means that you have to instantiate it in your script and now the dev need to know about obscure groovy script bindings? I do not understand what do we win by this?

**Comment by vickumar1981** [comment]
The other option, the way it is, requires you to pass the environment into the isDev() and isProd() functions.  Monadic functions are easier to understand, IMO.

**Comment by imuchnik** [comment]
This requires passing the variable on class creation. I do not see a strong argument for it . Either way, the variable is not available inside the class. I think that this requires more work for the developers, which kind of defeats the purpose. IF everyone feels very strongly about it, we can spend more time to figure out whether we can source that file into the scripts(as a script, rather than a class). That would achieve both goals.

**Comment by sleitner** [comment]
I think either way, it's prrretty clear what is going on, and either way, the user will have to look at the same groovy or (more likely) example code. I prefer this solution slightly because you pass fewer arguments and it's marginally clearer what each call is doing as a result (even though you have to make one extra call).

If we're only sourcing one variable from a script, would we really be gaining anything over just instantiating the class with $ENVIRONMENT?

**Comment by imuchnik** [comment]
@sleitner We pass exactly the same amount of arguments. We  gain just that- not having instantiate a class and still having it available. This has a cross-cutting nature, similarly to LOG function.

**Comment by vickumar1981** [comment]
@imuchnik   I don't see how it's different than JsJobBuilder or FlowBuilder where you have to pass arguments into the class creation.

def myJob = new JsJobBuilder(
        name: jobName,
        description: 'OAH front-end build',
        emails: ...

Why not just have a static method in JsJobBuilder called build job which takes all these variables as parameters?

So you'd have this:

class JsJobBuilder {
    ...
   static Job build(DslFactory factory, String name, String description, String email ...){
        ...
    }
}

EnvironmentUtils takes one parameter on construction right now, but that's why it's good b/c it's extensible.  In fact, static methods can be avoided altogether with a singleton design pattern :)

**Comment by vickumar1981** [comment]
So logger() is a good example.  That's a singleton usually.  In python, in java...always a singleton.

**Comment by sleitner** [comment]
"We pass exactly the same amount of arguments"

How is that? You pass your environment on every `isX` call. And it's very clear what instantiating a class means, whereas it's not as clear what passing an argument will do.

**Comment by imuchnik** [comment]
> In fact, static methods can be avoided altogether if you use a singleton design pattern :)

@vickumar1981  I don't see us enforcing a singleton pattern anywhere and we definitely do not need more then one instance of this class, which is also not true for builders. 

> So logger() is a good example. That's a singleton usually. In python, in java...always a singleton.

Note the absence of **new** keyword.

**Comment by imuchnik** [comment]
Also, where is the usage example and tests?

**Comment by vickumar1981** [comment]
you wouldn't need more than one instance of a builder either, if you pass all the arguments directly into the function...

build(DslFactory factory, String name, String description, String email ...){

Then I could do:

def job1 = JsJobBuilder.build (...)
def job2 = JsJobBuilder.build (...)

the problem is that arguments might get added on later, and that having a function call with 10 parameters is confusing.  So, we encapsulate them in a class.  I see the same thing with EnvironUtils, where we might want to add more settings later.  If EnvironUtils had 10 arguments instead of 1, then we wouldn't be making the case for using a static method.

**Comment by imuchnik** [comment]
@vickumar1981 I disagree,  we do need multiple instances of builders. 

I would argue against making this class do other things because we would be violating a "single responsibility principle". All this class should ever do is answer one question:"What environment am I in?"

**Comment by vickumar1981** [comment]
@imuchnik I can provide a decent use case - plugins.  If the list of plugins is different on prod vs. dev vs. staging, we might need to know about that.  A job could then, for example, ask EnvironUtils if that plugin is available for that environment.  

So I agree there's a single responsibility for the class, but it might, at some point later on, answer more than just a single question.  we might end up asking it a few more things about the environment down the road...about plugins and environment specific paths and stuff.

**Comment by imuchnik** [comment]
@vickumar1981 Sorry, I still do not see it.  Environment determination is a runtime state.  Plugin availability is an orthogonal concern that is not dependent on the environment, i.e this is something only Jenkins knows in any environment.

**Comment by sleitner** [comment]
@imuchnik How is plugin availability orthogonal to environment? Each environment has a Jenkins and each Jenkins has a set of plugins. They're perfectly correlated. We should probably unify the jenkinses, but for now, that's a fine example. And you can come up with any number of others: isDevIP(ip) for example.

That said, I think we're straying from the point a little, which is really that all of the static methods need to maintain a state, and we're asking the user to pass that state in over and over again, rather than maintaining it internally.

**Comment by imuchnik** [comment]
@sleitner orthogonal in way that plugins are not dependent on environments and vice versa. In simpler terms, knowing the environment does not guarantee the presence of the plugins and environments do not know anything about plugins present.   IsDevIp() will have live in some other utils.

Agreed about this conversation going off on the tangent. 

Perhaps seeing an example on code will help me see your thought process better?

**Comment by vickumar1981** [comment]
code just looks like this:

def env = new EnvironmentUtils (env: "${DEPLOY_ENV}")

and then:

if (env.isDev()) {
   ...//do something
}

if (env.isProd()) {
  ...//do something else
}

**Comment by imuchnik** [comment]
My problem is with new(). I don't think devs should have to do that in the script. 
Also that is not singleton pattern at all. If you are really opposed to the static class, but want a singleton we should do something like logger, where you would do Logger.getLogger()  to get a handle on the class instance.

**Comment by vickumar1981** [comment]
Yes.  I agree with that.  That's why I pointed out the singleton pattern.  IMO, EnvironmentUtils() should be a singleton.

Want me to make EnvironUtils a singleton() ?  Since that's something we all agree upon.

**Comment by imuchnik** [comment]
Yep, that would solve my issue with having developers new() up an instance .

**Comment by vickumar1981** [comment]
Done.  new way to invoke is like this:

def env = EnvironmentUtils.getInstance('dev')

**Comment by sleitner** [comment]
I'll leave it to Irina to plus one, but it's worth adding a usage example or comment.

**Comment by imuchnik** [comment]
:+1: but I am  not merging anything until @sleitner  does the same ;)

Also add example and this probably broke the test.

**Comment by vickumar1981** [comment]
fixed the broken test.

**Comment by sleitner** [comment]
woot! :+1:
